### PR TITLE
=str #17191 mapAsyncUnordered shall terminate after resume

### LINKED ISSUE #17191: FAILED: FlowMapAsyncUnorderedSpec
During validation of https://github.com/akka/akka/pull/17186

https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/2297/

```
[info] FlowMapAsyncUnorderedSpec:
Coroner Thread Count starts at 12 in akka.stream.scaladsl.FlowMapAsyncUnorderedSpec
[info] A Flow with mapAsyncUnordered
[info] - must produce future elements in the order they are ready (26 milliseconds)
[info] - must not run more futures than requested elements (732 milliseconds)
[info] - must signal future failure (17 milliseconds)
[WARN] [04/14/2015 00:17:32.218] [FlowMapAsyncUnorderedSpec-akka.actor.default-dispatcher-10] [akka://FlowMapAsyncUnorderedSpec/deadLetters] received dead letter from Actor[akka://FlowMapAsyncUnorderedSpec/deadLetters]: AsyncInput(MapAsyncUnordered(4,<function1>,<function1>),async(1),Success(1))
[WARN] [04/14/2015 00:17:32.219] [FlowMapAsyncUnorderedSpec-akka.actor.default-dispatcher-10] [akka://FlowMapAsyncUnorderedSpec/deadLetters] received dead letter from Actor[akka://FlowMapAsyncUnorderedSpec/deadLetters]: AsyncInput(MapAsyncUnordered(4,<function1>,<function1>),async(1),Success(4))
[WARN] [04/14/2015 00:17:32.219] [FlowMapAsyncUnorderedSpec-akka.actor.default-dispatcher-10] [akka://FlowMapAsyncUnorderedSpec/deadLetters] received dead letter from Actor[akka://FlowMapAsyncUnorderedSpec/deadLetters]: AsyncInput(MapAsyncUnordered(4,<function1>,<function1>),async(1),Success(2))
[info] - must signal error from mapAsyncUnordered (17 milliseconds)
[WARN] [04/14/2015 00:17:32.235] [FlowMapAsyncUnorderedSpec-akka.actor.default-dispatcher-2] [akka://FlowMapAsyncUnorderedSpec/user/$a/flow-4-2-mapAsyncUnordered] received dead letter from Actor[akka://FlowMapAsyncUnorderedSpec/user/$a/flow-4-2-mapAsyncUnordered#399808922]: AsyncInput(MapAsyncUnordered(4,<function1>,<function1>),async(1),Success(2))
[WARN] [04/14/2015 00:17:32.236] [FlowMapAsyncUnorderedSpec-akka.actor.default-dispatcher-2] [akka://FlowMapAsyncUnorderedSpec/user/$a/flow-4-2-mapAsyncUnordered] received dead letter from Actor[akka://FlowMapAsyncUnorderedSpec/user/$a/flow-4-2-mapAsyncUnordered#399808922]: AsyncInput(MapAsyncUnordered(4,<function1>,<function1>),async(1),Success(1))
[info] - must resume after future failure *** FAILED *** (3 seconds, 20 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) while expecting 5 messages (got 4)
[info]   at scala.Predef$.assert(Predef.scala:179)
[info]   at akka.testkit.TestKitBase$$anonfun$receiveN_internal$1.apply(TestKit.scala:633)
[info]   at akka.testkit.TestKitBase$$anonfun$receiveN_internal$1.apply(TestKit.scala:630)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:141)
[info]   at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
[info]   at scala.collection.AbstractTraversable.map(Traversable.scala:105)
[info]   at akka.testkit.TestKitBase$class.receiveN_internal(TestKit.scala:630)
[info]   at akka.testkit.TestKitBase$class.receiveN(TestKit.scala:621)
[info]   at akka.testkit.TestKit.receiveN(TestKit.scala:718)
[info]   at akka.stream.scaladsl.FlowMapAsyncUnorderedSpec$$anonfun$1$$anonfun$apply$mcV$sp$7.apply$mcV$sp(FlowMapAsyncUnorderedSpec.scala:125)
[info]   at akka.stream.scaladsl.FlowMapAsyncUnorderedSpec$$anonfun$1$$anonfun$apply$mcV$sp$7.apply(FlowMapAsyncUnorderedSpec.scala:112)
[info]   at akka.stream.scaladsl.FlowMapAsyncUnorderedSpec$$anonfun$1$$anonfun$apply$mcV$sp$7.apply(FlowMapAsyncUnorderedSpec.scala:112)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:953)
[info]   at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
[info]   at akka.stream.testkit.AkkaSpec.withFixture(AkkaSpec.scala:55)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:950)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:962)
[info]   at akka.stream.testkit.AkkaSpec.runTest(AkkaSpec.scala:55)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:390)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:427)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.WordSpecLike$class.runTests(WordSpecLike.scala:1021)
[info]   at akka.stream.testkit.AkkaSpec.runTests(AkkaSpec.scala:55)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1424)
[info]   at akka.stream.testkit.AkkaSpec.org$scalatest$WordSpecLike$$super$run(AkkaSpec.scala:55)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1067)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1067)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.WordSpecLike$class.run(WordSpecLike.scala:1067)
[info]   at akka.stream.testkit.AkkaSpec.org$scalatest$BeforeAndAfterAll$$super$run(AkkaSpec.scala:55)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:257)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:256)
[info]   at akka.stream.testkit.AkkaSpec.run(AkkaSpec.scala:55)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:466)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:677)
[info]   at sbt.ForkMain$Run.runTest(ForkMain.java:243)
[info]   at sbt.ForkMain$Run.runTestSafe(ForkMain.java:214)
[info]   at sbt.ForkMain$Run.runTests(ForkMain.java:190)
[info]   at sbt.ForkMain$Run.run(ForkMain.java:257)
[info]   at sbt.ForkMain.main(ForkMain.java:99)
```

### PR DESCRIPTION
**Pull request by rkuhn**
fixes #17191

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala [hunk 1]
```diff
@@ -171,6 +171,16 @@ class FlowMapAsyncSpec extends AkkaSpec {
       c.expectComplete()
     }
 
+    "finish after future failure" in {
+      import system.dispatcher
+      Await.result(Source(1 to 3).mapAsync(1, n ⇒ Future {
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala]: so the problem was when the last element throwed/resumed?

Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala]: only when the last one to come in throws, actually (which made this test racy in the unordered case)

**Comment by patriknw** [comment]
LGTM

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM

**Pull request merged**